### DISCUSSION TITLE
Allow Overrides and Implicit Interface Implementations to Override/Implement using Generically-Typed, Constrained members

### DISCUSSION THREAD

**Discussion by fitdev** [dotnet/csharplang#4113] [Language Ideas]
It would be nice if we could write code like this where generically-typed members (with appropriate constraints) can override members in the base class:

```cs
    public class ClassGenericDerived<T> : ClassBase where T : IFoo {
      //public override IFoo Foo { get; set; } //<- Works
      public override T Foo { get; set; } //<- Error: CS1715 'ClassGenericDerived<T>.Foo': type must be 'IFoo' to match overridden member 'ClassBase.Foo'
    }

    public class ClassBase {
      public virtual IFoo Foo { get; set; }
    }

    public interface IFoo { }
```

In the above code `T` is guaranteed to be of type `IFoo` due to generic constraint. So, why can't `T Foo {get; set; }` override base class member `public virtual IFoo Foo { get; set; }`?

Same logic would go for implicit interface implementations.

**Comment by canton7** [comment]
What happens if someone does this:

```cs
public class Foo1 : IFoo { }
public class Foo2 : IFoo { }

ClassBase c = new ClassGenericDerived<Foo1>();
c.Foo = new Foo2();
```

The signature of `ClassBase.Foo` says that someone can assign *any* type which implements `IFoo`. But your suggestion would mean that `ClassGenericDerived<Foo1>.Foo` is of type `Foo1`, so trying to assign any other implementation of `IFoo` wouldn't work. However that breaks the interface declared in `ClassBase`.

**Reply by HaloFour** [reply]
Agreed, this would make more sense if the property was `get`-only in which case it might qualify for covariant returns.  But I don't know if that applies to properties or to generic classes like this.

**Reply by canton7** [reply]
If a getter-only property doesn't come under covariant returns, I'd think it should be!

**Reply by HaloFour** [reply]
I believe it does based on this error from [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACBmAAhwCYCBBAgbwFgAoAhowgNzQRgFcIMCB7YAFZwAxjAIAFKgQDmcGAG4CAXzoradfEVIAhAiHJU6jJn2aIEaACZwiARgAMEqbIXLVQA===):

```
error CS8831: 'B.P': Target runtime doesn't support covariant types in overrides. Type must be 'object' to match overridden member 'A.P'
```

It also appears that the requested feature will also work with a getter-only property, based on the error message: [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACBmAAjQDt4EAzCAYzgIEkAxAeyYIG8CBfAWACh8COAEwFmrEPTHsufPgOEEAguz4E1gwgDc0CGAFcIGSSwIAFaQHM4MANwzePXnMIKAQgB4AKgD4CE5QDuABaItJ5+gkJIxqxsquoCTJqICGgAJmFmltZ2jpxAA==)

Someone running the preview can test for sure.  I wonder when SharpLab will support running on .NET 5.0.
